lib/OpenQA/Worker/Jobs.pm

```diff
@@ -357,22 +358,74 @@ sub _stop_job {
 
     if ($aborted eq "scheduler_abort") {
         log_debug('stop_job called by the scheduler. do not send logs');
-        _kill_worker($worker);
-        _reset_state;
+        _stop_job_announce(
+            $aborted, $job_id,
+            sub {
+                _kill_worker($worker);
+                _reset_state;
+            });
         return;
     }
 
-    # the update_status timers and such are gone by now (1st part), so we're
-    # basically "single threaded" and can block
-
-    my $status = {uploading => 1, worker_id => $workerid};
     api_call(
-        'post', "jobs/$job_id/status",
-        json     => {status => $status},
-        callback => sub { _stop_job_2($aborted, $job_id, $host); });
+        post => "jobs/$job_id/status",
+        json => {
+            status => {
+                uploading => 1,
+                worker_id => $workerid,
+            },
+        },
+        callback => sub {
+            _stop_job_announce(
+                $aborted, $job_id,
+                sub {
+                    _stop_job_kill_and_upload($aborted, $job_id, $host);
+                });
+        },
+    );
+}
+
+sub _stop_job_announce {
+    my ($aborted, $job_id, $callback) = @_;
+
+    my $ua      = Mojo::UserAgent->new(request_timeout => 10);
+    my $job_url = $job->{URL};
+    return $callback->() unless $job_url;
+
+    try {
+        my $url = "$job_url/broadcast";
+        my $tx  = $ua->build_tx(
+            POST => $url,
+            json => {
+                stopping_test_execution => $aborted,
+            },
+        );
+
+        log_info('trying to stop job gracefully by announcing it to command server via ' . $url);
+        my $mojo_loop_running = Mojo::IOLoop->is_running;
+        $ua->start(
+            $tx,
+            sub {
+                my ($ua_from_callback, $tx) = @_;
+                my $keep_ref_to_ua = $ua;
+                my $res            = $tx->res;
+
+                if (!$res->is_success) {
+                    log_error('unable to stop the command server gracefully: ');
+                    log_error($res->code ? $res->to_string : 'command server likely not reachable at all');
+                }
+                $callback->();
+            });
+    }
+    catch {
+        log_error('unable to stop the command server gracefully: ' . $_);
+
+        # ensure stopping is proceeded (failing announcement is not critical)
+        $callback->();
+    };
```

Comment:
Since I'm not 100% sure no errors will occur here I added the try-catch. Although it is quite unlikely, the worker will get stuck in the error case otherwise. I really want to avoid this by all means.

Reply:
The `$ua->build_tx` and `$ua->start` don't normally throw exceptions. You'd have to pass them the wrong arguments for that.